Never regard macro rules with compile_error! invocations as unused

est31 opened the pull request:
The very point of compile_error! is to never be reached, and one of
the use cases of the macro, currently also listed as examples in the
documentation of compile_error, is to create nicer errors for wrong
macro invocations. Thus, we should never warn about unused macro arms
that contain invocations of compile_error.

See also https://github.com/rust-lang/rust/pull/96150#issuecomment-1126599107 and the discussion after that.

Furthermore, the PR also contains two commits to silence `unused_macro_rules` when a macro has an invalid rule, and to add a test that `unused_macros` does not behave badly in the same situation.

r? @petrochenkov as I've talked to them about this

eggyal commented on compiler/rustc_expand/src/mbe/macro_rules.rs:
```diff
@@ -651,6 +660,30 @@ fn check_matcher(sess: &ParseSess, def: &ast::Item, matcher: &[mbe::TokenTree])
+    match rhs {
+        Delimited(_sp, d) => {
+            let has_compile_error = d.tts.array_windows::<3>().any(|[ident, bang, args]| {
+                if let Token(ident) = ident &&
```
I'm not sure a simple scan of the token tree is sufficient here.  Suppose a macro rule produces conditional output:

```rust
macro_rules! foo {
    (one) => {
        #[cfg(sometimes)]
        compile_error!();

        #[cfg(not(sometimes))]
        // no compile_error!
        todo!();
    };
}
```

Then shouldn't we still be linting the macro as unused when `not(sometimes)`?

est31 replied:
Unfortunately it's hard to do "static" analysis for macros. Macros act in a heavily untyped domain of token trees, and there are many more ways the `compile_error` macro could be invoked. See also: https://github.com/rust-lang/rust/pull/96150#issuecomment-1132816837

petrochenkov replied:
It's a lint, it can use heuristics.
If detecting the sequence "`compile_error` `!`" removes more false positives than introduces false negatives, then it's fine.

petrochenkov commented on compiler/rustc_expand/src/mbe/macro_rules.rs:
```diff
@@ -542,8 +542,17 @@ pub fn compile_declarative_macro(
     // We only take the span of the lhs here,
     // so that the spans of created warnings are smaller.
+    // Also, we are only interested in non-foreign macros.
     let rule_spans = if def.id != DUMMY_NODE_ID {
```
```suggestion
    let rule_spans = if valid && def.id != DUMMY_NODE_ID {
```
This is off-topic, but we don't need to mark the arms as unused for malformed macros, this would only result in additional diagnostic noise.

est31 replied:
Oh good point currently this:

```rust
#![deny(unused_macro_rules)]

macro_rules! foo {
    (v) => {};
    () => 0;
}
fn main() {
    foo!(v);
}
```

gives:

```
error: macro rhs must be delimited
 --> src/main.rs:5:11
  |
5 |     () => 0;
  |           ^

error: 1st rule of macro `foo` is never used
 --> src/main.rs:4:5
  |
4 |     (v) => {};
  |     ^^^
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![deny(unused_macro_rules)]
  |         ^^^^^^^^^^^^^^^^^^

error: 2nd rule of macro `foo` is never used
 --> src/main.rs:5:5
  |
5 |     () => 0;
  |     ^^

```

which is very wrong, because the first arm is very much used. For some reason, the unused macros lint does not have this issue, as in this does lint:

```Rust
#![deny(unused_macro_rules)]

macro_rules! foo {
    (v) => {};
    () => 0;
}

fn main() {}
```
```
error: macro rhs must be delimited
 --> src/main.rs:5:11
  |
5 |     () => 0;
  |           ^

warning: unused macro definition: `foo`
 --> src/main.rs:3:14
  |
3 | macro_rules! foo {
  |              ^^^
  |
  = note: `#[warn(unused_macros)]` on by default
```

But the unused_macros lint is gone if you use `foo`. I guess it might be an idea to add a test for it.

petrochenkov commented on compiler/rustc_expand/src/mbe/macro_rules.rs:
```diff
@@ -651,6 +660,30 @@ fn check_matcher(sess: &ParseSess, def: &ast::Item, matcher: &[mbe::TokenTree])
+fn has_compile_error_macro(rhs: &mbe::TokenTree) -> bool {
+    use mbe::TokenTree::*;
+    match rhs {
+        Delimited(_sp, d) => {
```
```suggestion
        TokenTree::Delimited(_sp, d) => {
```
Nit: could you avoid the glob import above?

est31 commented:
As requested, I've removed the glob import. I've also added two commits. The first added commit silences `unused_macro_rules` when a macro has a malformed rule. The second added commit ensures that the `unused_macros` lint works correctly when a macro has a malformed rule. I don't think it needs to be turned off as well, but feel free to disagree.

r? @petrochenkov

bors commented:
:pushpin: Commit 787e24cdfd5bef5136540287dd918e48b8a645db has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 787e24cdfd5bef5136540287dd918e48b8a645db -->
<!-- homu: {"type":"Approved","sha":"787e24cdfd5bef5136540287dd918e48b8a645db","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 787e24cdfd5bef5136540287dd918e48b8a645db with merge fa68e73e9947be8ffc5b3b46d899e4953a44e7e9...
<!-- homu: {"type":"BuildStarted","head_sha":"787e24cdfd5bef5136540287dd918e48b8a645db","merge_sha":"fa68e73e9947be8ffc5b3b46d899e4953a44e7e9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6842902933?check_suite_focus=true)
Approved by: petrochenkov
Pushing fa68e73e9947be8ffc5b3b46d899e4953a44e7e9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6842902933?check_suite_focus=true"},"merge_sha":"fa68e73e9947be8ffc5b3b46d899e4953a44e7e9"} -->

rust-timer commented:
Finished benchmarking commit (fa68e73e9947be8ffc5b3b46d899e4953a44e7e9): [comparison url](https://perf.rust-lang.org/compare.html?start=7092d42c04a56378d2e748f98169b728b1f1d5d1&end=fa68e73e9947be8ffc5b3b46d899e4953a44e7e9&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=7092d42c04a56378d2e748f98169b728b1f1d5d1&end=fa68e73e9947be8ffc5b3b46d899e4953a44e7e9&stat=instructions:u)

- Primary benchmarks: no relevant changes found
- Secondary benchmarks: 😿 relevant regression found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | 0.3%     | 0.3% | 1         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | N/A      | N/A | 0         |
| All 😿🎉 (primary) | N/A      | N/A | 0         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=7092d42c04a56378d2e748f98169b728b1f1d5d1&end=fa68e73e9947be8ffc5b3b46d899e4953a44e7e9&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: 😿 relevant regression found
- Secondary benchmarks: mixed results

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 3.2%     | 3.2% | 1         |
| Regressions 😿 <br /> (secondary) | 2.4%     | 2.6% | 4         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -3.3%    | -3.3% | 1         |
| All 😿🎉 (primary) | 3.2%     | 3.2% | 1         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=7092d42c04a56378d2e748f98169b728b1f1d5d1&end=fa68e73e9947be8ffc5b3b46d899e4953a44e7e9&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: no relevant changes found
- Secondary benchmarks: 🎉 relevant improvement found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -2.9%    | -2.9% | 1         |
| All 😿🎉 (primary) | N/A      | N/A | 0         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
